Automattic/jetpack
# Jetpack Boost: Improve DataSync error handling

## Proposed changes:
This lays the groundwork for improved error management in DataSync.

Previously it was really difficult diagnosing where DataSync errors come from because schema parsers had no idea of what's going on and when they're being parsed int he call tree.

In addition to that - it was impossible to communicate to the front end by just throwing errors from DataSync entries.

This introduces two new error classes:

`Schema_Internal_Error` for internal parsing errors - they will be caught within DataSync parsing process and converted into `Schema_Parsing_Error` augmenting the object with a bit of data along the way so that it's possible to trace exactly what went wrong, what was expected and what was passed. 

DataSync schemas are also now JSON Serializable, meaning that if you `json_encode` a schema, you get back a JSON representation of the schema. This could possibly be used for additional debugging utilities like Zod vs Backend schema validator, but that's for a later date.

For right now, here's a brief overview what's changed:

* Added Error classes
* Every type is now serializable and inspectable as a string
* Added a lot of unit tests for error handling
* JSON Serializable Schemas



Here's a list of things that still aren't done, but now that the groundwork is there and stable, I'll be able to come up with follow-up PRs:

* Full Backend to Zod error handling compatibility. Throw errors in DataSync backend, and gracefully receive them in `useDataSync()`
* "Debug Mode" - while this PR contains a single "error_log" that's going to log validation errors with a few more details - this is still not done. We'll want more "dev mode safety", like fallback value validation when WP_DEBUG is enabled (because currently our fallback values are incorrect, like expecting an array but returning null, etc. )
* Proper error handling at the endpoint level. 
* Documentation

Ideas for the future:
JSON Deserializing - if there's enough time left, I wonder if we can replace `Schema::as_somehting()` with a PHP array to also deserialize it. It's actually not impossible at this point to create a CLI tool that reads a Zod schema and creates a PHP array and ensures that they're in sync.

### 🦋 Side-quest
As I was figuring errors out, I accidentally went on a side quest, replacing all `json_decode( $value )` calls with `json_decode( $value, ARRAY_A )`.

Anonymous objects in PHP aren't great:

```
wp> $obj = new stdClass();
=> object(stdClass)#8529 (0) {
}
wp> empty($obj);
=> bool(false)
wp> !!$obj
=> bool(true)
wp> $arr = array();
=> array(0) {
}
wp> empty($arr);
=> bool(true)
wp> !!$arr
=> bool(false)
wp>
```

This can produce bugs in case an object isn't handled properly. 

```
wp> $x = json_decode("{}");
=> object(stdClass)#8531 (0) {
}
wp> empty($x);
=> bool(false)
wp> $y = json_decode("{}", ARRAY_A);
=> array(0) {
}
wp> empty($y);
=> bool(true)
wp>
```


### Other information:


## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
1. Apply patch
2. Build everything
3. No really, build everything
4. Test that everything works as expected in Boost Plugin
5. Dump schemas and produce errors - have a look at the logs and enjoy that you finally don't have to hunt for which key/schema has failed.
